Library/Homebrew/download_strategy.rb

```diff
@@ -340,29 +342,29 @@ def parse_basename(url, search_query: true)
         end
       end
 
-      if uri.query && search_query
-        "#{uri.path}?#{uri.query}"
-      else
-        uri.path
+      if (uri_path = uri.path)
+        components[:path] = uri_path.split("/").map do |part|
+          URI::DEFAULT_PARSER.unescape(part).presence
+        end.compact
+      end
+
+      if search_query && (uri_query = uri.query)
+        components[:query] = URI.decode_www_form(uri_query).map(&:second)
       end
     else
-      url
+      components[:path] = [url]
     end
 
-    uri_path = URI.decode_www_form_component(uri_path)
-    query_regex = /[^?&]+/
-
     # We need a Pathname because we've monkeypatched extname to support double
     # extensions (e.g. tar.gz).
     # Given a URL like https://example.com/download.php?file=foo-1.0.tar.gz
     # the basename we want is "foo-1.0.tar.gz", not "download.php".
-    Pathname.new(uri_path).ascend do |path|
-      ext = path.extname[query_regex]
-      return path.basename.to_s[/#{query_regex.source}#{Regexp.escape(ext)}/] if ext
+    components.values.flatten.reverse_each do |path|
+      path = Pathname(path)
+      return path.basename.to_s if path.extname.present?
     end
 
-    # Strip query string
-    File.basename(uri_path)[query_regex]
+    File.basename(components[:path].last || "")
```

Comment:
```suggestion
    File.basename(components[:path].last.presence || "")
```

although I wonder if we should just return `""` rather than running `File.basename` on an empty string?

Reply:
Probably - just requires splitting into a couple lines.